### ISSUE TITLE
interruptionEventStream is not working on iOS

### ISSUE THREAD

**Issue by BringTheChill** [ryanheise/audio_session#28]
Hi. For me, interruptionEventStream is not working anymore on iOS and I do not know why (I use audioplayers package). With another words, if my audio is interrupted nothing happens. :( 
On my initState() after super.initState() I have this function **_Provider.of<Data>(context, listen: false).audioInterruptedListener();_**

And this is the function: 

```dart
  void audioInterruptedListener() {
    AudioSession.instance.then(
      (audioSession) async {
        bool playInterrupted = false;
        bool narratorInterrupted = false;
        await audioSession.configure(AudioSessionConfiguration.speech());
        audioSession.interruptionEventStream.listen(
          (event) {
            if (event.begin) {
              switch (event.type) {
                case AudioInterruptionType.duck:
                case AudioInterruptionType.pause:
                case AudioInterruptionType.unknown:
                  if (playInterrupted) {
                    playerBackground.resume();
                    if (playerState == AudioPlayerState.PAUSED &&
                        narratorInterrupted) {
                      playerNarrator.resume();
                      narratorInterrupted = false;
                    }
                    playInterrupted = false;
                    break;
                  }
                  if (!playInterrupted) {
                    playInterrupted = true;
                    playerBackground.pause();
                    if (playerState == AudioPlayerState.PLAYING) {
                      playerNarrator.pause();
                      narratorInterrupted = true;
                    }
                  }
                  break;
              }
            } else {
              switch (event.type) {
                case AudioInterruptionType.duck:
                case AudioInterruptionType.pause:
                  if (playInterrupted) {
                    playerBackground.resume();
                    if (playerState == AudioPlayerState.PAUSED &&
                        narratorInterrupted) {
                      playerNarrator.resume();
                      narratorInterrupted = false;
                    }
                    playInterrupted = false;
                  }
                  break;
                case AudioInterruptionType.unknown:
                // The interruption ended but we should not resume.
              }
            }
          },
        );
      },
    );
  }
```

**Comment by ryanheise** [comment]
The problem could be in one of 3 places.

Was there a change in your audio_session version, or a change in audioplayers, or a change in your code?

**Comment by BringTheChill** [comment]
@ryanheise There was more than a change. I updated audio_session and audioplayeres to the latest version. Code not changed.

**Comment by ryanheise** [comment]
Are you able to downgrade audio_session and test?

**Comment by BringTheChill** [comment]
@ryanheise I did it and it's working (with unsound null safety). So what should I do? I need a null safety version tho so I can release my app to beta testers and after this for production.

**Comment by ryanheise** [comment]
I'd like to fix it (with your cooperation if you can), are you able to provide a minimal reproduction project that I can clone and run? E.g. By modifying the example to reproduce the bug?

**Comment by paurakhsharma** [comment]
Can confirm, We upgrated from version `audio_session: ^0.0.11` to `audio_session: ^0.1.0` and facing same issue on iOS.

**Comment by ryanheise** [comment]
Thank you for the reproduction project. Just to confirm if I understand correctly, your example works if I downgrade the dependencies? What specifically were you testing as the source of the audio interruptions?

**Comment by BringTheChill** [comment]
> your example works if I downgrade the dependencies?

 yes
> What specifically were you testing as the source of the audio interruptions?

 Siri voice (I hope I understood your question)

**Comment by paurakhsharma** [comment]
> What specifically were you testing as the source of the audio interruptions?

Spotify, just as #27

**Comment by ryanheise** [comment]
This is quite puzzling. I have compared the diffs between 0.0.11 and 0.1.0 and the logic is pretty much the same.

May I ask if you are seeing this log message appear from the objective C code?

```objc
    NSLog(@"audioInterrupt");
```

Objective C logs might not appear in the usual Flutter console, so you may need to run this through Xcode to see it.

If that log is happening, then at least we know the even is being received on the platform side and is just getting lost in communication to the Flutter side.

**Comment by paurakhsharma** [comment]
I use the session config for `speech`
with
```
session.configure(AudioSessionConfiguration.speech())
```

These are the logs I see:
1) In vscode
  i) When Spotify takes the focus of the audio `audioInterrupt` log is displayed in vscode console. My audio player is still playing but is ducked and cannot be heard.
  ii) When I pause Spotify I get the log `routeChange detected` and then  `audioInterrupt`. Myaudio player doesn't get the volume back (i.e it is playing but cannot be heard)


2) In Xcode
  Same output. Same result.

Hope it helps.

**Comment by BringTheChill** [comment]
> This is quite puzzling. I have compared the diffs between 0.0.11 and 0.1.0 and the logic is pretty much the same.
> 
> May I ask if you are seeing this log message appear from the objective C code?
> 
> ```objc
>     NSLog(@"audioInterrupt");
> ```
> 
> Objective C logs might not appear in the usual Flutter console, so you may need to run this through Xcode to see it.
> 
> If that log is happening, then at least we know the even is being received on the platform side and is just getting lost in communication to the Flutter side.

@ryanheise yes I got "audioInterrupt" on my console. I said this in a comment above too.

**Comment by ryanheise** [comment]
Ah, yes I missed that you mentioned that. This is I suppose good news because it means the event is being detected at the low levels and it's just failing to make its way up to the higher levels.

There is an intermediate level in the `AVAudioSession` class. If you listen to the interruption stream directly from this class, do you receive the notifications? i.e.

```dart
    AVAudioSession().interruptionNotificationStream.listen((note) {
      print('received interrupt');
      print('type: ${note.type}');
      print('wasSuspended: ${note.wasSuspended}');
    });
```

**Comment by paurakhsharma** [comment]
@ryanheise 
We inserted the hooks after audio_session initialization.

```dart
 final session = await AudioSession.instance;
 await session.configure(AudioSessionConfiguration.speech());

 AVAudioSession().interruptionNotificationStream.listen((note) {
   print('received interrupt');
   print('type: ${note.type}');
   print('wasSuspended: ${note.wasSuspended}');
 });

```

We even closed the app and re-run it after making this change.

But we couldn't see this log output.

Here is the log we got instead:

When the audio player starts
```
routeChange detected
```


When I open and play Spotify
```
audioInterrupt
```

When I pause Spotify
```
routeChange detected
audioInterrupt
```

When I go back to the app
```
dnssd_clientstub read_all(11) DEFUNCT
[VERBOSE-2:FlutterObservatoryPublisher.mm(143)] Could not register as server for FlutterObservatoryPublisher. Check your network settings and relaunch the application.
dnssd_clientstub read_all(20) DEFUNCT
[VERBOSE-2:FlutterObservatoryPublisher.mm(143)] Could not register as server for FlutterObservatoryPublisher. Check your network settings and relaunch the application.
dnssd_clientstub read_all(31) DEFUNCT
[VERBOSE-2:FlutterObservatoryPublisher.mm(143)] Could not register as server for FlutterObservatoryPublisher. Check your network settings and relaunch the application.
dnssd_clientstub read_all(33) DEFUNCT
[VERBOSE-2:FlutterObservatoryPublisher.mm(143)] Could not register as server for FlutterObservatoryPublisher. Check your network settings and relaunch the application.
BackgroundSession <13A7D914-C21A-43B1-AB9D-E60C4C437399> connection to background transfer daemon interrupted
BackgroundSession <13A7D914-C21A-43B1-AB9D-E60C4C437399> connection to background transfer daemon invalidated
BackgroundSession <FD822123-5BED-435D-ADE3-DEE72D3E996D> connection to background transfer daemon interrupted
BackgroundSession <FD822123-5BED-435D-ADE3-DEE72D3E996D> connection to background transfer daemon invalidated
BackgroundSession <B6674567-56BF-48E1-89F9-FC4EFFB3CDC1> connection to background transfer daemon interrupted
BackgroundSession <B6674567-56BF-48E1-89F9-FC4EFFB3CDC1> connection to background transfer daemon invalidated

```

**Comment by BringTheChill** [comment]
I fixed this issue replacing this package with assets_audio_player. Thank you for your support! I will let this issue open in case @paurakhsharma wants to continue with this package.

**Comment by ryanheise** [comment]
Yes, please do leave it open until the issue is resolved.

**Comment by ryanheise** [comment]
OK, I think I may see the issue. Let me try a fix.

**Comment by paurakhsharma** [comment]
Thanks, @ryanheise will let you know shortly.

**Comment by paurakhsharma** [comment]
@ryanheise Can confirm this is fixed for me. Thank you so much 🎉 🚀

**Comment by ryanheise** [comment]
Great! Thanks for helping to test this. I'll put out a new release now.

**Comment by paurakhsharma** [comment]
Thank you for putting your time and effort to make our life easier 🙂

**Comment by ryanheise** [comment]
The fix has now been published as release `0.1.1`.

Note, the next release will probably be `0.2.0` which will move to the new `rxdart`.
